src/tree/tree_model.cc

```diff
@@ -17,158 +22,601 @@ DMLC_REGISTER_PARAMETER(TreeParam);
-      } else {
-        fo << ",gain=" << std::setprecision(float_max_precision) << tree.Stat(nid).loss_chg
-           << ",cover=" << std::setprecision(float_max_precision) << tree.Stat(nid).sum_hess;
+    // This turns out to be tricky, as `dmlc::Parameter::Load(JSONReader*)` doesn't
+    // support loading nested json objects.
+    if (kwargs.find("condition_node_params") != kwargs.cend()) {
+      auto const& cnp = kwargs["condition_node_params"];
+      for (auto const& kv : cnp) {
+        param_.condition_node_params += kv.first + '=' + "\"" + kv.second + "\" ";
       }
+      kwargs.erase("condition_node_params");
     }
-    if (format == "json") {
-      fo << ", \"children\": [";
-    } else {
-      fo << '\n';
+    if (kwargs.find("leaf_node_params") != kwargs.cend()) {
+      auto const& lnp = kwargs["leaf_node_params"];
+      for (auto const& kv : lnp) {
+        param_.leaf_node_params += kv.first + '=' + "\"" + kv.second + "\" ";
+      }
+      kwargs.erase("leaf_node_params");
+    }
+
+    if (kwargs.find("edge") != kwargs.cend()) {
+      if (kwargs["edge"].find("yes_color") != kwargs["edge"].cend()) {
+        param_.yes_color = kwargs["edge"]["yes_color"];
+      }
+      if (kwargs["edge"].find("no_color") != kwargs["edge"].cend()) {
+        param_.no_color = kwargs["edge"]["no_color"];
+      }
+      kwargs.erase("edge");
     }
-    DumpRegTree(fo, tree, fmap, tree[nid].LeftChild(), depth + 1, false, with_stats, format);
-    DumpRegTree(fo, tree, fmap, tree[nid].RightChild(), depth + 1, true, with_stats, format);
-    if (format == "json") {
-      fo << std::endl;
-      for (int i = 0; i < depth + 1; ++i) {
-        fo << "  ";
+    auto const& extra = kwargs["graph_attrs"];
+    static std::string const kGraphTemplate = "    graph [ {key}=\"{value}\" ]\n";
+    for (auto const& kv : extra) {
+      param_.graph_attrs += SuperT::Match(kGraphTemplate,
+                                     {{"{key}", kv.first},
+                                      {"{value}", kv.second}});
+    }
+
+    kwargs.erase("graph_attrs");
+    if (kwargs.size() != 0) {
+      std::stringstream ss;
+      ss << "The following parameters for graphviz are not recognized:\n";
+      for (auto kv : kwargs) {
+        ss << kv.first << ", ";
       }
-      fo << "]}";
+      LOG(WARNING) << ss.str();
     }
   }
-}
+
+ protected:
+  // Only indicator is different, so we combine all different node
+  // types into this function.
+  std::string PlainNode(RegTree const& tree, int32_t nid, uint32_t depth) override {
+    auto split = tree[nid].SplitIndex();
+    auto cond = tree[nid].SplitCond();
+    static std::string const kNodeTemplate = "    {nid} [ label=\"{label}\" {params}]\n";
+    static std::string const kLabelTemplate = R"({fname}{<}{cond})";
+
+    // Indicator only has fname.
+    bool has_less = split > fmap_.Size() || (split < fmap_.Size() && fmap_.type(split));
```

Comment:
We should also disallow the case where `split` is equal to `fmap_.Size()`.
```suggestion
    bool has_less = split >= fmap_.Size() || (split < fmap_.Size() && fmap_.type(split));
```